data/src/nonempty/map.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright © 2019-2020 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{
+    borrow::Borrow as _,
+    collections::{BTreeMap, HashMap},
+    convert::TryFrom,
+    hash::{BuildHasher, Hash},
+    marker::PhantomData,
+    ops::Deref,
+};
+
+/// Alias for a [`NonEmptyMap`] backed by a [`BTreeMap`]
+pub type NonEmptyOrderedMap<K, V> = NonEmptyMap<K, V, BTreeMap<K, V>>;
+
+/// Alias for a [`NonEmptyMap`] backed by a [`HashMap`]
+pub type NonEmptyHashMap<K, V, S> = NonEmptyMap<K, V, HashMap<K, V, S>>;
+
+/// Simple newtype wrapper around a map type which witnesses that the map
+/// contains at least one element.
+///
+/// Non-mutating methods of the underlying map are available through the
+/// [`Deref`] impl. Additionally, mutating methods which cannot leave an empty
+/// map behind are provided. To construct a [`NonEmptyMap`], use the [`TryFrom`]
+/// impl.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct NonEmptyMap<K, V, Inner = BTreeMap<K, V>> {
+    inner: Inner,
+    _marker: PhantomData<(K, V)>,
+}
+
+impl<K: Ord, V> NonEmptyOrderedMap<K, V> {
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, [`None`] is returned. If the
+    /// map did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be == without being identical.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        self.inner.insert(key, value)
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&K, &mut V)> {
+        self.inner.iter_mut()
+    }
+
+    pub fn values_mut(&mut self) -> impl Iterator<Item = &mut V> {
+        self.inner.values_mut()
+    }
+}
+
+impl<K, V, S> NonEmptyHashMap<K, V, S>
+where
+    K: Eq + Hash,
+    S: BuildHasher,
+{
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, [`None`] is returned. If the
+    /// map did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be == without being identical.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        self.inner.insert(key, value)
+    }
+
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&K, &mut V)> {
+        self.inner.iter_mut()
+    }
+
+    pub fn values_mut(&mut self) -> impl Iterator<Item = &mut V> {
+        self.inner.values_mut()
+    }
+}
+
+impl<K, V> TryFrom<BTreeMap<K, V>> for NonEmptyMap<K, V> {
+    type Error = ();
```

Comment:
Is having an empty error type going to lead to confusion? I know it's pretty obvious that there's only really one way to fail to construct a non empty collection, but if you're looking at a trace from a program that's just crashed I feel like it will be much quicker to figure out what's going wrong if the error explicitly says that you tried to construct a non empty collection without values rather than having no message.

Reply:
Ya. I think it should be a contructor returning `Option<Self>`, really. Wait for reroll.